Loop geomspace interface

Frank-Duffy opened the pull request:
Added geomspace interface to LoopTask.

Takes start, stop and num values from user to create geomspace np array for loop values in Exopy.

Frank-Duffy committed:
added necessary files for geomspace interface and updated declarations

Frank-Duffy committed:
updated rounding for generated values

Frank-Duffy committed:
commented out enaml coverage plugin install

MatthieuDartiailh commented:
Looks like you disabled too many installs... I hope to push a fix for enaml_coverage_plugin tomorrow. This way we will be able to run as usual.

MatthieuDartiailh commented:
The release for enaml_coverage_plugin is done so you can revert your last commit

Frank-Duffy commented:
Thanks will do

On Thu, Mar 9, 2023 at 10:32 AM Matthieu Dartiailh ***@***.***>
wrote:

> The release for enaml_coverage_plugin is done so you can revert your last
> commit
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Exopy/exopy/pull/182#issuecomment-1462258855>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATYTR2KJGDSZW7Q67HT5VRTW3HZXPANCNFSM6AAAAAAVMPLHX4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Frank-Duffy committed:
Revert "commented out enaml coverage plugin install"

This reverts commit 7d8043d60dd22ae3d4d6ef57f69179df69efb78d.

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_geomspace_interface.py:
```diff
@@ -0,0 +1,84 @@
+
+
+class GeomspaceLoopInterface(TaskInterface):
+    """ Common logic for all loop tasks.
```
Comment looks wrong

Frank-Duffy replied:
Just updated

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_geomspace_interface.py:
```diff
@@ -0,0 +1,84 @@
+    stop = Str('100.0').tag(pref=True, feval=Feval(types=numbers.Real))
+
+    #: Step between loop values.
+    num = Str('10').tag(pref=True, feval=Feval(types=numbers.Real))
```
The validation should enforce we get an integer

Frank-Duffy replied:
just updated

Frank-Duffy replied:
So I changed it to Str, but then that broke the format_and_eval_string call later on. I set it up as a Str originally because that allows the num variable to be set by a database entry for another task.

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_geomspace_interface.py:
```diff
@@ -0,0 +1,84 @@
+        num = task.format_and_eval_string(self.num)
+
+        # Make sure the sign of the step makes sense.
+        num = -abs(num) if start > stop else abs(num)
```
Since you use a number of points this does not seem relevant

Frank-Duffy replied:
just removed this

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_geomspace_interface.py:
```diff
@@ -0,0 +1,84 @@
+        stop_digit = abs(Decimal(str(stop)).as_tuple().exponent)
+        start_digit = abs(Decimal(str(start)).as_tuple().exponent)
+        digit = max((start_digit, stop_digit))
+        stop = round(stop, digit)
```
Why do you only round the stop ?

Frank-Duffy replied:
Update: I removed line 74 because the following lines of code provide the rounding for the iterable array.

MatthieuDartiailh reviewed (commented):
I left a bunch of comments, and ideally, I would like to see some tests.

Frank-Duffy committed:
updated geomspace interface comment

Frank-Duffy committed:
enforce num to be integer

Frank-Duffy committed:
removed sign check on num

Frank-Duffy commented:
> I left a bunch of comments, and ideally, I would like to see some tests.

Hi and thanks for the feedback. Which tests would be best for this?

MatthieuDartiailh commented:
I would mostly adapt the tests existing for Linspace.

Frank-Duffy committed:
Changed num back to Str

Int broke type completion selection of database entries on the user interface.

MatthieuDartiailh commented:
About your last commit you could change the validation to enforce an integer rather than just a real number.

Frank-Duffy committed:
enforced num to be an integer

Frank-Duffy committed:
added geomspace_interface fixture

Frank-Duffy committed:
added geomspace creation check to geomspace interface

Frank-Duffy committed:
added space so that I could leave a comment on the PR

Is there a way to comment on code outside of the commit but in the changed file in the PR without having to do this?

MatthieuDartiailh commented:
Please ping me when you want another review, your update to the tests do not do anything yet.

Frank-Duffy commented:
Will do.

On Mon, Apr 24, 2023 at 3:21 PM Matthieu Dartiailh ***@***.***>
wrote:

> Please ping me when you want another review, your update to the tests do
> not do anything yet.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Exopy/exopy/pull/182#issuecomment-1520704739>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATYTR2NACYXIDWHFKBR3HMLXC3HCLANCNFSM6AAAAAAVMPLHX4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Frank-Duffy committed:
added num value check

added try except block for checking that the num value provided is greater than or equal to 1. num is the number of points to be created in the geomspace array.

Frank-Duffy committed:
added geomspace rounding test

added a test that checks if the values of the generated array are rounded as intended.

MatthieuDartiailh commented:
I cleanup up a bunch of things in #180 which should address your issue. Could you rebase and test again ?

Frank-Duffy committed:
updated geomspace tests

Frank-Duffy committed:
added unit test for geomspace num sign during check

Frank-Duffy committed:
added unit test for exception handling during check

Frank-Duffy committed:
updated comments

Frank-Duffy commented:
> Please ping me when you want another review, your update to the tests do not do anything yet.

I just added a test to ensure that exceptions building the geomspace array are caught during the check sequence of the program. Would you mind taking a look and letting me know if I should add/modify anything?

All the best,
Frank

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_geomspace_interface.py:
```diff
@@ -0,0 +1,89 @@
+            test = False
+            mess = 'Loop task did not succeed to create a geomspace array: {}'
+            traceback[err_path + '-geomspace'] = mess.format(e)
+
+        
```
I believe here you should set the `values` database entries (after rounding)

MatthieuDartiailh commented on exopy/tasks/tasks/logic/views/loop_geomspace_view.enaml:
```diff
@@ -0,0 +1,58 @@
+                entries_updater << \
+                    interface.task.list_accessible_database_entries
+                tool_tip = EVALUATER_TOOLTIP
+    SplitItem:
```
Missing blank line above

MatthieuDartiailh reviewed (changes requested):
Bunch of things I would like to see fixed or with at least a comment. And I will check coverage when the report comes out.

Frank-Duffy committed:
fixed inadvertent changes to ci.yml

Frank-Duffy committed:
added blank line to the end of declarations.enaml

Frank-Duffy committed:
added blank line between SplitItems in loop_gemosapce_view.enaml

Frank-Duffy committed:
removed redundant code

loop_task.py already ensures that the start value is written into the value entry.

Frank-Duffy committed:
added generate_geomspace_array helper func and assoc tests

Frank-Duffy committed:
changed test name

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_geomspace_interface.py:
```diff
@@ -44,47 +44,44 @@ def check(self, *args, **kwargs):
+        array = np.fromiter((round(value, digit)
                                 for value in raw_values),
                                np.float64, len(raw_values))
-        task.write_in_database('loop_values', np.array(iterable))
```
You need to write in the database

MatthieuDartiailh commented on exopy/tasks/tasks/logic/loop_task.py:
```diff
@@ -133,7 +133,7 @@ def _perform_loop_timing(self, iterable):
         self.write_in_database('point_number', len(iterable))
         self.write_in_database('loop_values', np.array(iterable))
-
+        
```
un-stripped whitespaces

MatthieuDartiailh replied:
VS code can be configured to remove those for you on save

MatthieuDartiailh commented on tests/tasks/tasks/logic/test_loop_task.py:
```diff
@@ -92,12 +92,12 @@ def test_geomspace_handling_of_rounding(monkeypatch, geomspace_interface):
+    generated = geomspace_interface.generate_geomspace_array()
     expected = np.array([0.01, 0.02, 0.03, 0.05, 0.08, 0.13,
                          0.22, 0.36, 0.6, 1.])
-    np.testing.assert_array_equal(lt.database_entries['iterable'], expected)
```
Testing with the database is a good idea

MatthieuDartiailh commented:
You are missing a test exercising perform and some tests handling validation errors in checks.

Frank-Duffy committed:
added write to database for loop_values

Frank-Duffy committed:
fixed un-stripped white space

Frank-Duffy committed:
moved write in database to generator helper func

Frank-Duffy committed:
updated generate func and its test and added perform test

Frank-Duffy committed:
added unit test for try/except block in check method

Frank-Duffy committed:
updated exception test intent

Frank-Duffy committed:
corrected test comment

MatthieuDartiailh commented:
Sorry for the once again slow response. This is looking good. I would just like to see a test exercising that a bad start/stop/step is properly caught. Do not hesitate to ping me as needed.

Frank-Duffy commented:
No problem and thank you!

I will ping you when I have another update.

All the best,
Frank

On Wed, Aug 30, 2023 at 3:12 PM Matthieu Dartiailh ***@***.***>
wrote:

> Sorry for the once again slow response. This is looking good. I would just
> like to see a test exercising that a bad start/stop/step is properly
> caught. Do not hesitate to ping me as needed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Exopy/exopy/pull/182#issuecomment-1699701386>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATYTR2ICZWX5RMZIX35BG5TXX6GA7ANCNFSM6AAAAAAVMPLHX4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Frank-Duffy committed:
added tests to exercise user input exception handling

MatthieuDartiailh commented:
Great ! Could you add an entry to the changelog (CHANGES) and I will merge

Frank-Duffy committed:
Update CHANGES

Frank-Duffy commented:
> Great ! Could you add an entry to the changelog (CHANGES) and I will merge

Just added an entry and pushed the commit.

Frank-Duffy commented:
> Great ! Could you add an entry to the changelog (CHANGES) and I will merge

Actually I just noticed that there's one line that isn't being executed by the unit tests and it's driving the codecov down. I'm going to write something really quick to exercise that line. It's the write_in_database under the try: of the check().

Frank-Duffy committed:
Fixed test function names
